### PR TITLE
Use tags for passing runner metadata

### PR DESCRIPTION
**Pull request by dgholz**
for fac/dev-platform#607

Prep work for switching to fleets, `CreateFleet` doesn't allow userdata to be set so it has to come from the launch template, and per-run settings will have to be passed in a different way. I'm trying tags.

### CODE REVIEW

File: src/aws.js [hunk 1]
```diff
@@ -1,41 +1,54 @@
...
+  // push returns new size of array, so don't use its result as the function return value
+  return userdata_prefix.concat(
+    'INSTANCE_ID=$(cat /var/lib/cloud/data/instance-id)',
+    'TOKEN_HEADER="X-aws-ec2-metadata-token: $(curl -sX PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")"',
+    'AWS_DEFAULT_REGION="$(curl -sH "$TOKEN_HEADER" http://169.254.169.254/latest/meta-data/placement/availability-zone | sed -e"s/[a-z]*\$//")"',
+    'export AWS_DEFAULT_REGION',
+    // yikes
+    // avoid `aws ec2 ... | while read`, since the pipeline starts a subshell for the `while`
+    // so variables set in the loop disappear when it's done
+    'while read -r key value; do declare "RUNNER_$(echo -n "$key" | tr \'[:lower:]\' \'[:upper:]\' | tr -cs \'[:alnum:]\' _ )"="$value"; done < <(aws ec2 describe-tags --output text --filter Name=resource-id,Values="$INSTANCE_ID" --query "Tags[?starts_with(Key, \\`Ec2GithubRunner:\\`)].[Key, Value]" --output text | cut -f2- -d:)',
```
Comment (dgholz) [inline src/aws.js:32]: This is pretty hard to read & test! I think we should not have a bunch of Bash in a Javascript action, and this should get moved to a helper script that's baked into the image, or a lambda.

File: src/aws.js [hunk 2]
```diff
@@ -1,41 +1,54 @@
...
-      './run.sh',
-    ];
+    );
   }
+  // push returns new size of array, so don't use its result as the function return value
+  return userdata_prefix.concat(
+    'INSTANCE_ID=$(cat /var/lib/cloud/data/instance-id)',
+    'TOKEN_HEADER="X-aws-ec2-metadata-token: $(curl -sX PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")"',
+    'AWS_DEFAULT_REGION="$(curl -sH "$TOKEN_HEADER" http://169.254.169.254/latest/meta-data/placement/availability-zone | sed -e"s/[a-z]*\$//")"',
+    'export AWS_DEFAULT_REGION',
```
Comment (singhprd) [inline src/aws.js:28]: Do I not know bash, or does this need a `$AWS_DEFAULT_REGION`?

File: src/aws.js (hunk 1 above)
Comment (singhprd) [reply src/aws.js:32]: Yep we've talked about this, something we'd both like to refactor later on.

**Review by singhprd — APPROVED** [review]
✅

File: src/aws.js (hunk 1 above)
Comment (stevenwilliamson) [reply src/aws.js:32]: +1 when/if we refactor vote for the simple option of bundling a script utility in the AMI

Comment (dgholz) [reply src/aws.js:32]: Yeah, it should be a simple enough utility, that looks for tags with a prefix (`Ec2GithubRunner`, set in the userdata) and spits out the vars (either sets them directly or returns an `eval` string)

File: src/aws.js [hunk 3]
```diff
@@ -1,41 +1,54 @@
...
+    'export RUNNER_ALLOW_RUNASROOT DOTNET_SYSTEM_GLOBALIZATION_INVARIANT',
+    `./config.sh --url "$RUNNER_URL" --token ${githubRegistrationToken} --name "$INSTANCE_ID" --labels "$RUNNER_LABEL"`,
+    './run.sh',
+  );
+}
+
+function makeMetadataTags(label) {
+    const { owner, repo } = github.context.repo;
+    const url = `${github.context.serverUrl}/${owner}/${repo}`
+    return [ { Key: "Ec2GithubRunner:Url", Value: url }, { Key: "Ec2GithubRunner:Label", Value: label }, ];
```
Comment (stevenwilliamson) [inline src/aws.js:44]: There are some generic guidelines for tagging in notion here fwiw https://www.notion.so/freeagent/AWS-Tagging-Policy-a752ee992cbe4e49af97a417f93de750

TL;DR it's best to opt for lower-case to avoid typos and tags not been found down the line because someone got the casing incorrect. Hyphens for word separation and any custom ones specifc to fa prefixed with freeagent: to avoid name space collision issues on tags.

File: src/aws.js (hunk 1 above)
Comment (dgholz) [reply src/aws.js:32]: Peter moved it to the AMI in fac/github-runners-ami-image#16
